[incubator/crate] Add CrateDB

OJFord opened the pull request:
#### What this PR does / why we need it:

Adds a chart for [CrateDB](https://crate.io).

> CrateDB is a distributed SQL database built on top of a NoSQL foundation. It combines the familiarity of SQL with the scalability and data flexibility of NoSQL, enabling developers to:
>
>    * Use SQL to process any type of data, structured or unstructured
>    * Perform SQL queries at realtime speed, even JOINs and aggregates
>    * Scale simply
>
> Customers often use CrateDB to store and query machine data. This is because CrateDB makes it easy and economical to handle the velocity, volume, and diversity of machine and log data. In fact, customers have reported CrateDB ingesting millions of data points per second, while also querying terabytes of data in real time… 20x faster than their previous database and on 75% less database hardware.

#### Which issue this PR fixes
No issue.

#### Special notes for your reviewer:
Previous PR from @taliaga closed due to inactivity: #8432.

The same commits are present here, plus my bump to the latest Crate version.

@autophagy is a Crate maintainer and expressed interest in crate/crate#7904, and provided feedback in the original PR (which as far as I can tell @taliaga already actioned).

#### Checklist

- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

taliaga committed:
Add crateDB chart

Signed-off-by: Tomas Aliaga <tomas.aliaga@gmail.com>

taliaga committed:
Add mantainers and suggestions from PR 8432

Signed-off-by: Tomas Aliaga <tomas.aliaga@gmail.com>

taliaga committed:
Reduce initialdelayseconds in livenessProbe

Signed-off-by: Tomas Aliaga <tomas.aliaga@gmail.com>

OJFord committed:
Bump to Crate v4.0.3

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Remove non-Helm org owners, add requested approver

Signed-off-by: Ollie Ford <dev@ojford.com>

chaudum commented on incubator/crate/Chart.yaml:
```diff
@@ -0,0 +1,14 @@
+name: crate
+version: 0.1.0
+description: CrateDB is a distributed SQL database that handles massive amounts of machine data in real-time. This is a Helm chart for CrateDB, to simplify even further its deployment on Kubernetes clusters.
+appVersion: 4.0.3
```
The current stable version is already `4.0.4`.

chaudum replied:
In my opinion it would make sense to set the `appVersion` value to `4.0`, which is the minor version of CrateDB which is definitely compatible with the provided settings. So you can be sure any version between 4.0.0 and 4.0.n is compatible with this Helm chart.

OJFord replied:
That was just to match the image (which was the latest at the time) - I like your idea; I don't think it goes against Helm's 'best practices' which says on it:

> It is a way of specifying the version of the application. For example, the drupal chart may have an appVersion: 8.2.1, indicating that the version of Drupal included in the chart (by default) is 8.2.1.

Taking 'may' per RFC2119 and given that saying 'this includes 4.0 by default' if it includes '4.0.4' is hardly false, I think we can do that.

```suggestion
appVersion: 4.0
```

chaudum commented on incubator/crate/NOTES.txt:
````diff
@@ -0,0 +1,9 @@
+kubectl port-forward crate-0 4200:4200
+```
+
+Now open an explorer at http://localhost:4200, you should see the CrateDB's Admin UI.
````
```suggestion
Now open a browser at http://localhost:4200, you should see the CrateDB's Admin UI.
```

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| crate.clusterName | Name of CrateDB cluster. See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#id3) | my-cluster |
+| crate.crateHeapSize | See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/environment.html) | 1g |
+| crate.recoverAfterNodes | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 1 |
+| crate.replicas | The number of instances to deploy. Also the number of expected nodes in the cluster formation | 1 |
```
In CrateDB the term "replica" has a different meaning. Even though I know that you mean the K8s term replica to indicate the number of pods the Statefulset should spawn, it could still cause confusion for people working with CrateDB.
May I suggest to name the configuration parameter `crate.numberOfNodes` to clearly indicate is has to do with CrateDB instances, and not shard replicas.

chaudum replied:
Optionally you could name the parameter `statefulset.replicas`, since you're also using the `service` prefix.

OJFord replied:
Thanks, I like `numberOfNodes` for the symmetry with `recoverAfterNodes`.

```suggestion
| crate.numberOfNodes | The number of instances to deploy. Also the number of expected nodes in the cluster formation | 1 |
```

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| --------- | ----------- | ------- |
+| crate.clusterName | Name of CrateDB cluster. See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#id3) | my-cluster |
+| crate.crateHeapSize | See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/environment.html) | 1g |
+| crate.recoverAfterNodes | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 1 |
```
This gateway setting should be greater equals the number of nodes in the CrateDB cluster (`x = n // 2 + 1`, where `n` is the number of expected nodes)

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| crate.crateHeapSize | See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/environment.html) | 1g |
+| crate.recoverAfterNodes | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 1 |
+| crate.replicas | The number of instances to deploy. Also the number of expected nodes in the cluster formation | 1 |
+| crate.recoverAfterTime | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 5m |
```
gateway.recover_after_time has the default value `5m`. IMO is is not really worth making it configurable

chaudum replied:
Also, link to the guide is broken:

```suggestion
| crate.recoverAfterTime | See [docs](https://crate.io/docs/crate/guide/en/latest/scaling/multi-node-setup.html#gateway-configuration) | 5m |
```

OJFord replied:
I'm all for smaller and simpler to start with at least.

```suggestion
```

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+
+| Parameter | Description | Default |
+| --------- | ----------- | ------- |
+| crate.clusterName | Name of CrateDB cluster. See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#id3) | my-cluster |
```
The link to the guide is broken:

```suggestion
| crate.clusterName | Name of CrateDB cluster. See [docs](https://crate.io/docs/crate/guide/en/latest/scaling/multi-node-setup.html#id3) | my-cluster |
```

OJFord replied:
That version matches the default in the chart, IMO it shouldn't inadvertently link to 5.0 docs while providing a 4.x image.

OJFord replied:
Ah I see what you mean *guide* only has latest.

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| --------- | ----------- | ------- |
+| crate.clusterName | Name of CrateDB cluster. See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#id3) | my-cluster |
+| crate.crateHeapSize | See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/environment.html) | 1g |
+| crate.recoverAfterNodes | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 1 |
```
Link to the guide is broken:

```suggestion
| crate.recoverAfterNodes | See [docs](https://crate.io/docs/crate/guide/en/latest/scaling/multi-node-setup.html#gateway-configuration) | 1 |
```

OJFord replied:
(As above)

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| crate.recoverAfterTime | See [docs](https://crate.io/docs/crate/guide/en/4.0.3/scaling/multi-node-setup.html#gateway-configuration) | 5m |
+| http.cors.enabled | Whether CORS is enabled. See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/node.html#cross-origin-resource-sharing-cors) | False |
+| http.cors.allowOrigin | CORS origin to allow (if enabled). See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/node.html#cross-origin-resource-sharing-cors) | * |
+| image.name | Container image name | crate |
```
I am questioning if the image name should be configurable, because it is not for nothing a CrateDB Helm chart :)

OJFord replied:
Seems reasonable to me, let's have only the tag (i.e. crate version) configurable.

```suggestion
```

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+| http.cors.allowOrigin | CORS origin to allow (if enabled). See [docs](https://crate.io/docs/crate/reference/en/4.0.3/config/node.html#cross-origin-resource-sharing-cors) | * |
+| image.name | Container image name | crate |
+| image.tag | Container image tag (version) | 4.0.3 |
+| image.pullPolicy | Container pull policy | IfNotPresent |
```
If only official CrateDB Docker images are used, tags will never change their content, so there is no need for changing the imagePullPolicy to something else than its default value `IfNotPresent`.

OJFord replied:
```suggestion
```

Per https://github.com/helm/charts/pull/16318/files#r321134391

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+
+## Configuration
+
+The following table lists the configurable parameters of the crateDB chart and their default values.
```
```suggestion
The following table lists the configurable parameters of the CrateDB chart and their default values.
```

chaudum commented on incubator/crate/values.yaml:
```diff
@@ -0,0 +1,53 @@
+  clusterPort: 4300
+  # https://crate.io/docs/crate/reference/en/latest/interfaces/postgres.html
+  postgresName: postgres
+  postgresPort: 5432
```
May I suggest a change in the naming of the service ports to:

```yaml
service:
  ports:
    http: 4200  # HTTP port which serves the Admin UI as well as acts as SQL endpoint for clients using the HTTP protocol.
    transport: 4300  # this should very likely not be configurable, because it is not a user-facing port but only used for transport between CrateDB nodes
    psql: 5432  # Postgres port for clients using the Postgres Protocol
```

What do you think?

OJFord replied:
Agreed, done. Only reason I can think of for configuring `transport` differently would be because for some reason the UI/psql port is really wanted to be 4300... but that's mad, as above let's keep it simpler and more prescriptive for now at least.

chaudum commented on incubator/crate/values.yaml:
```diff
@@ -0,0 +1,53 @@
+  requests:
+    memory: 1Gi
+  limits:
+    memory: 2Gi
```
What about also adding CPU request/limit?

OJFord replied:
Hm, what about actually doing something with the memory request/limit while we're at it...!

chaudum replied:
Resource limits are rendered into the template [here](https://github.com/helm/charts/pull/16318/files#diff-cf4b8dda181f1a7229191f17ca16be93R58) if that is what you mean?

chaudum replied:
Other than that, I think `requests` and `limits` could (and should) have the same value, alternatively only provide `limits`.
The memory limit should be at least 2 times and up to 3 times higher than the allocated heap size for CrateDB. https://crate.io/docs/crate/guide/en/latest/performance/memory.html

I don't know whether it would somehow be possible to configure the required resource memory limit based on the CrateDB heap size parameter (or vice versa).

chaudum replied:
Ah, I just noticed that there is a comment already regarding the heap configuration [here](https://github.com/helm/charts/pull/16318/files#diff-2708716e3dd4d331124ecac79a27ba6dR20).

OJFord replied:
Ah - so it is there, bad grep, thanks.

Yes, we can do that. How about having configurable `crateHeapSize` (as it is) and `limits.memory`, but inferring the `requests.memory` as 2x heap size? And the _default_ `limits.memory` at 2x as well, or 3x?

chaudum commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+          value: {{ .Values.crate.clusterName | quote }}
+        - name: RECOVER_AFTER_NODES
+          value: {{ .Values.crate.recoverAfterNodes | quote }}
+        livenessProbe:
```
The discovery requires to know the name of the CrateDB nodes, so we use the pod name as node name (usage see comment above):

```suggestion
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        livenessProbe:
```

chaudum commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+          - -Clicense.enterprise=false
+          - -Ccluster.name={{ .Values.crate.clusterName }}
+          - -Cnetwork.host="0.0.0.0"
+          - -Cdiscover.seed_providers=srv
```
The configuration is missing the `cluster.initial_master_nodes` setting.

```suggestion
          - -Cdiscover.seed_providers=srv
          - -Cnode.name=#{POD_NAME}
          - -Ccluster.initial_master_nodes={{ range $i, $sequence :=  (seq .Values.crate.replicas) }}crate-{{ $i }},{{ end}}
```

`POD_NAME` is obtains from the metadata.name, see comment below.

PS: I am not familiar with the Go templating, so the loop may not be accurate. It should output the names of the pod, separated by comma, e.g.  if there are 3 replicas in this Statefulset, the line should read like so:

```yaml
          - -Ccluster.initial_master_nodes=crate-0,crate-1,crate-2
```

Where `crate` is the name of the Pod (defined by `metadata.name`.

chaudum commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+          - -Chttp.cors.allow-origin={{ .Values.http.cors.allowOrigin }}
+          - -Cpath.data=/data
+          {{- if .Values.resources.limits.cpu }}
+          - -Cprocessors={{ .Values.resources.limits.cpu }}
```
Would you mind showing the number of CPUs in the `values.yaml`?

OJFord replied:
I don't think that's the right place or name for this value - I haven't used the Crate parameter but it sounds like it means `N` physical count? A k8s `limits` value of `1` would be a single hyperthread or vCPU, but it's often fractional, e.g. `900m` or `1.5` - would Crate accept that?

(Not to mention that the *limit* my not be what is actually allocated, in fact it probably isn't if it != `requests`, because this is happening on startup, not under load.)

chaudum replied:
In CrateDB `processors` is used to calculate the size of threadpools. On bare metal or VMs the value is derived from the system, however inside Docker this value is not correct, because it reports the number of CPUs of the underlying host. That's why it must/should be overwritten when CrateDB is run in Docker.
As you stated correctly, it should be a positive integer, and cannot handle fractional values.

OJFord replied:
I _suppose_ the correct value then is the ceil of `requests`, e.g. `900m -> 1`, `1.5 -> 2`. But I'm not really sure that's how it works, or whether `1.5` vCPUs as a k8s unit can be a share of any (larger than one) number of underlying vCPUs = hyperthreads.

Is underestimating, e.g. using `--Cprocessors=2 # ceil 1.5` when we actually have a 0.5 share of three vCPUs an issue?

chaudum replied:
We also use `ceil` when we deploy our own clusters and have not seen any problems yet. However we usually also only use full CPU core values.

chaudum commented on incubator/crate/README.md:
```diff
@@ -0,0 +1,36 @@
+Install this chart from this folder using `helm install .`.
+You may want to adjust first your preferred number of nodes in `values.yml`.
+
+## Configuration
```
Is it intentional, that the README does not document all available parameters? If so, would it make sense to keep somewhere a list of all parameters?

OJFord replied:
I think this should be that list, but as discussed elsewhere if there's no reason to be configurable items shouldn't be in that list. When we've settled on a `values.yaml` I'll go through and make sure they're all documented here.

chaudum reviewed (commented):
Hi @OJFord 

Thanks for your effort to get a CrateDB Helm chart into the official repository. I am engineer at Crate.io, so I felt free to comment on the PR.

I left a few comments on how the chart could be improved. The most important thing is the missing setting for master node election, without which a cluster would not form ([`cluster.initial_master_nodes` guide](https://crate.io/docs/crate/guide/en/latest/scaling/multi-node-setup.html#master-node-election)).

If you need more help and background information to certain configuration settings, please feel free to ask any time.

OJFord commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+          - -Cdiscover.seed_providers=srv
+          - -Cdiscovery.srv.query=_{{ .Values.service.clusterName }}._tcp.{{ .Values.service.name }}.{{ .Release.Namespace }}.svc.cluster.local
+          - -Cgateway.recover_after_nodes={{ .Values.crate.recoverAfterNodes }}
+          - -Cgateway.recover_after_time={{ .Values.crate.recoverAfterTime }}
```
```suggestion
```

Config removed https://github.com/helm/charts/pull/16318#discussion_r321130566

OJFord commented on incubator/crate/values.yaml:
```diff
@@ -0,0 +1,53 @@
+  crateHeapSize: 1g
+  recoverAfterNodes: 1
+  replicas: 1  # If you want more replicas, checkout the links above.
+  recoverAfterTime: 5m
```
```suggestion
```

Config removed https://github.com/helm/charts/pull/16318#discussion_r321130566

OJFord commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+          privileged: true
+      containers:
+      - name: crate
+        image: {{ .Values.image.name }}:{{ .Values.image.tag }}
```
```suggestion
        image: crate:{{ .Values.image.tag }}
```

Per https://github.com/helm/charts/pull/16318/files#r321134391

OJFord commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,98 @@
+      containers:
+      - name: crate
+        image: {{ .Values.image.name }}:{{ .Values.image.tag }}
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
```
```suggestion
        imagePullPolicy: IfNotPresent
```

Per https://github.com/helm/charts/pull/16318/files#r321134391

OJFord committed:
Simplify configuration

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Fix typos

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Bump to Crate v4.0.4

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Fix missing initial_master_nodes for 4.x

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Infer recoverAfterNodes default from numberOfNodes

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Fix -Cprocessors integer value

Must be integer, and not greater than available threads.
https://crate.io/docs/crate/reference/en/latest/config/node.html#cpu

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Infer requests.memory and default limit from heapSize

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,115 @@
+          valueFrom:
+            resourceFieldRef:
+              resource: requests.cpu
+              divisor: 1
```
This might still be fractional, I'm struggling to parse the docs, need to verify.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#quantity-resource-core

OJFord replied:
Nope, this works great. I tested on a deployment with `requests/cpu: 128m`, and this `resourceFieldRef` evaluates to `1`, which is exactly what we want for Crate's `-Cprocessors` flag.

OJFord committed:
Default crate.clusterName to value of app

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Set docs links to match appVersion (no patch)

Less hassle to maintain, and they redirect to unpatched minor anyway.

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Fix docs guide links

Only ther reference docs have versioned URLs, guide is latest-only.

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Document all configuration in one place

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Make array formatting consistent

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Fix syntax when PV not enabled

Signed-off-by: Ollie Ford <dev@ojford.com>

OJFord committed:
Add test values

Signed-off-by: Ollie Ford <dev@ojford.com>

fabianbaier commented on incubator/crate/templates/statefulset.yaml:
```diff
@@ -0,0 +1,116 @@
+                {{ if $i }},{{ end }}crate-{{ $i }}
+           {{- end }}
+            - -Ccluster.name={{ include "crate.clusterName" . }}
+            - -Cdiscover.seed_providers=srv
```
```suggestion
            - -Cdiscovery.seed_providers=srv
```

OJFord replied:
Thanks! 🤦‍♂
